File: heap/maximum_product_of_two_elements_1464.py
import heapq

# largest or smallest 나오면 heap을 생각하면 됨 
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        
        heap = []

        for i in range(len(nums)-1):
            for j in range(i+1, len(nums)):
                cur_product = (nums[i]-1) * (nums[j]-1)
                heapq.heappush(heap, -cur_product)
        
        return -heapq.heappop(heap)

File: 4_heap/maximum_product_of_two_elements_1464.cpp
#include <queue>
#include <iostream>

using namespace std;

class Solution {
    public:
        int maxProduct(vector<int>& nums) {
            
            //default = max_heap, less<int> 
            priority_queue<int, vector<int>, less<int>> pq;


            for (int i = 0 ; i < nums.size() -1 ; i++){
                for (int j = i+1; j < nums.size() ; j++){
                    int cur_product = (nums[i] -1 )*(nums[j] -1);
                    pq.push(cur_product);
                }
            }

            return pq.top();
        }
};